### PR TITLE
Fix `dir` method for windows shell sessions

### PR DESCRIPTION
**Pull request by pingport80**
## Summary
This PR fixes a bug in the `dir` method under `msf/core/post/file.rb`.

This method is expected to return a list of entries(files/directories) for the specified directory but for the `windows` platform(non-meterpreter) which also includes session type `powershell`, it seems inconsistent because it returns the output of `C:\Users> dir [directory]` command. Also the output for the `dir` command is different in case of `powershell` so they both should be parsed differently.

## Before
```ruby
>> dir '.'
=> 
[" Volume in drive Z is Computer",
 " Volume Serial Number is 23dsf92309",
 " Directory of Z:\\Users\\user\\Links",
 "06-06-2021  08.18 PM    <DIR>          .",
 "06-06-2021  08.18 PM    <DIR>          ..",
 "16-10-2020  01.48 PM               532 Desktop.lnk",
 "16-10-2020  01.48 PM               977 Downloads.lnk",
 "14-04-2021  12.27 PM             7,168 windows_shell_stage_64.exe",
 "12-04-2021  04.12 PM           250,368 windows_stage_tcp.exe",
 "               4 File(s)        259,045 bytes",
 "               2 Dir(s)  217,593,282,560 bytes free"]
>> 
```
## After
```ruby
>> dir '.'
=> ["Desktop.lnk", "Downloads.lnk", "windows_shell_stage_64.exe", "windows_stage_tcp.exe"]
>> 
```
## Note
I've not tested this for powershell session type because I was facing some issues while running the `reverse_powershell` payload.

### CODE REVIEW

File: lib/msf/core/post/file.rb [hunk 1]
```diff
@@ -73,8 +73,14 @@ def dir(directory)
       return session.fs.dir.entries(directory)
     end
 
+    if session.type == 'powershell'
+      dir = session.shell_command_token("Get-ChildItem \"#{directory}\" | Format-Table Name").split(/[\r\n]+/)
+      dir.slice!(0..2)
```
Comment (bwatters-r7) [inline lib/msf/core/post/file.rb:78]: There's a paranoid part of me that wants to check the size before we slice.

Comment (pingport80) [reply lib/msf/core/post/file.rb:78]: Cool, will add a check!

**Comment by bwatters-r7** [comment]
Hi there!  I wanted to let you know that I have not forgotten about this; while testing it last week, I saw some fundamental issues with how cmd_exec worked in shell sessions, and I tried to work up a fix to that first.  I'm hoping that once https://github.com/rapid7/metasploit-framework/pull/15362 gets landed, I can take another swing at this PR so we can run it against the built-in test modules.

**Comment by pingport80** [comment]
> Hi there! I wanted to let you know that I have not forgotten about this; while testing it last week, I saw some fundamental issues with how cmd_exec worked in shell sessions, and I tried to work up a fix to that first. I'm hoping that once #15362 gets landed, I can take another swing at this PR so we can run it against the built-in test modules.

sure, no worries! :slightly_smiling_face:

**Comment by pingport80** [comment]
I would like to propose a `powershell` label for this : )

File: lib/msf/core/post/file.rb [hunk 2]
```diff
@@ -53,7 +53,7 @@ def pwd
       if session.platform == 'windows'
         # XXX: %CD% only exists on XP and newer, figure something out for NT4
         # and 2k
-        return session.shell_command_token("echo %CD%")
+        return session.shell_command_token("echo %CD%").to_s.strip
```
Comment (bwatters-r7) [inline lib/msf/core/post/file.rb:56]: Curious if we need this after the changes in https://github.com/rapid7/metasploit-framework/pull/15362 landed (Specifically these three lines: https://github.com/rapid7/metasploit-framework/pull/15362/files#diff-d173cbcdf8260e163d18837d5154c57e0646ad954187bce96fae8e78105cc67fR66) where we normalize the line endings already.

Comment (bwatters-r7) [reply lib/msf/core/post/file.rb:56]: To be clear, I'm not super passionate about removing it, as it is sort of a belt/suspenders thing, and it does not hurt anything having it there.

Comment (pingport80) [reply lib/msf/core/post/file.rb:56]: Hey @bwatters-r7, I just tested this and looks like we still need to strip it.
```
>> pwd
=> "C:\\Users\\User\\Links\r\n"
```

File: lib/msf/core/post/file.rb [hunk 3]
```diff
@@ -73,8 +73,14 @@ def dir(directory)
...
 
+    if session.type == 'powershell'
+      dir = session.shell_command_token("Get-ChildItem \"#{directory}\" | Format-Table Name").split(/[\r\n]+/)
+      dir.slice!(0..2) if dir.length > 2
+      return dir
+    end
+
     if session.platform == 'windows'
-      return session.shell_command_token("dir #{directory}").split(/[\r\n]+/)
+      return session.shell_command_token("dir /b \"#{directory}\"").split(/[\r\n]+/)
```
Comment (bwatters-r7) [inline lib/msf/core/post/file.rb:83]: You probably want to protect this split from nil:
`/home/tmoose/rapid7/metasploit-framework/lib/msf/core/post/file.rb:77:in `dir': undefined method `split' for nil:NilClass (NoMethodError)`

Comment (pingport80) [reply lib/msf/core/post/file.rb]: Done :)

**Comment by bwatters-r7** [comment]
# Release Notes
This PR ensures that the shell `dir` command returns a list.

**Pull request merged**